### DISCUSSION TITLE
WiFi connection

### DISCUSSION THREAD

**Discussion by ssombra** [arkypita/LaserGRBL#1147] [Q&A]
I am trying to connect Lasergrbl via wifi so I got one of those Uno + wifi boards:

![image](https://user-images.githubusercontent.com/42944253/97064881-1d2d6700-156f-11eb-8388-22cba1c71133.png)

I followed the instructions in the [Lasergrbl webpage for wifi connection](http://lasergrbl.com/usage/wifi-with-esp8266/): I installed the libraries and the boards definitions, I can upload GRBL to the Uno side of the board, and it seems that I can also upload the sketch to the ESP8266 side of the board:

![image](https://user-images.githubusercontent.com/42944253/97065081-924d6c00-1570-11eb-8b10-bdeb0f5b31ab.png)

After that, I reboot the board and I can see the ESP8266 wifi but everytime I try to connect to it I get an error message that I can't connect to that network. Sometimes I can't see the ESP8266 network. 

![Screenshot 2020-10-23 212250](https://user-images.githubusercontent.com/42944253/97065891-4ef5fc00-1576-11eb-99f6-138c1b5c2129.jpg)

I tried with the telnet sketch, with the Websocket sketch and I even tried with the Laserweb sketch and always the same problem. I think that it is a problem with the libraries. Are the instructions still valid for the current board definitions and libraries? should I try with older versions? anyone with this board was succesful?

**Comment by gmmanonymus111** [comment]
I had exactly that board for a while. I found a telnet bridge library (I can't recall it where), that was working okayish (I had to lower status updates, and some processed gcodes stuck in LaserGRBL's buffer, the connection sometimes just dropped because esp buffer got full). After this, I found [ESP-Link](https://github.com/jeelabs/esp-link) which is similar to the telnet bridge, but easier to config, more stable, and supports multiple connections at the same time. Flashing it is a bit harder, because you have to use esptool command manually, but that is documented in their wiki.

**Comment by ssombra** [comment]
> esptool



> I had exactly that board for a while. I found a telnet bridge library (I can't recall it where), that was working okayish (I had to lower status updates, and some processed gcodes stuck in LaserGRBL's buffer, the connection sometimes just dropped because esp buffer got full). After this, I found [ESP-Link](https://github.com/jeelabs/esp-link) which is similar to the telnet bridge, but easier to config, more stable, and supports multiple connections at the same time. Flashing it is a bit harder, because you have to use esptool command manually, but that is documented in their wiki.

Thanks for the reply. Will try to take a look at the documentation, but this seems way beyond my understanding.

**Comment by gmmanonymus111** [comment]
If you need a step-by-step tutorial, how to flash it, I'll try to write an easy one, it might be useful for other users

**Comment by ssombra** [comment]
> If you need a step-by-step tutorial, how to flash it, I'll try to write an easy one, it might be useful for other users

That would be great, thanks for the offer.

**Comment by arkypita** [comment]
I have exactly the same board. I use it without any problem with the telnet sketch, since 2017.
I haven't needed to recompile it since I installed it, so I don't know if the libraries it uses are still compatible.

However, the sketch itself does not nothing else that read from wifi and write to serial, and vice versa.
All the WiFi stuff is managed by the libraries.

**Comment by ssombra** [comment]
> I have exactly the same board. I use it without any problem with the telnet sketch, since 2017.
> 
> I haven't needed to recompile it since I installed it, so I don't know if the libraries it uses are still compatible.
> 
> 
> 
> However, the sketch itself does not nothing else that read from wifi and write to serial, and vice versa.
> 
> All the WiFi stuff is managed by the libraries.

I am sure my problem is not with the sketch but the libraries as I tried 3 different sketches. Maybe it is the way I upload the sketch to the ESP8266. Maybe you don't remember now, but I do the following after adding the libraries to the ARDUINO ide: 
1) put switches 5, 6 and 7 in the "on" position and upload the sketch to the ESP8266 using the ARDUINO IDE
2) Then I reboot the board with all switch positions I can imagine 
3) sometimes The ESP wifi is not visible and the times they are, I get the can't connect message when trying to connect to it 

Is there something I am missing? Should I need to do something else to the ESP?

**Comment by ssombra** [comment]
> I had exactly that board for a while. I found a telnet bridge library (I can't recall it where), that was working okayish (I had to lower status updates, and some processed gcodes stuck in LaserGRBL's buffer, the connection sometimes just dropped because esp buffer got full). After this, I found [ESP-Link](https://github.com/jeelabs/esp-link) which is similar to the telnet bridge, but easier to config, more stable, and supports multiple connections at the same time. Flashing it is a bit harder, because you have to use esptool command manually, but that is documented in their wiki.

I was able to flash ESP-Link following the instructions in the site and using ESPTOOL, but I still have the same problem, no Access Point at all or, in the best case, it appears but it is not possible to connect to it. I have actually 2 of these boards and I have the problem with both. So still no luck for adding wifi to my Laser engraver with these boards.

**Comment by gmmanonymus111** [comment]
It's sad to hear you are still facing this issue. Could you try erasing all data on your board? (`esptool.py --port {YOURUSBPORT} erase_flash`) There might be a corrupted setting there, and by default, this region is ignored when flashing new firmware.

**Comment by arkypita** [comment]
Yesterday I spent the whole evening trying to program an esp-01 with the sketch, but it seems that I forgot how to do it because I was unable to upload the sketch to the chip!

I'll try again in the next few days

**Comment by ssombra** [comment]
> It's sad to hear you are still facing this issue. Could you try erasing all data on your board? (`esptool.py --port {YOURUSBPORT} erase_flash`) There might be a corrupted setting there, and by default, this region is ignored when flashing new firmware.

I tried this but the problem persists. I think I will give up with this board and get some cheap ESP-01s modules. It is a shame because I really liked the idea of having all integrated in one board.

**Comment by ssombra** [comment]
> Yesterday I spent the whole evening trying to program an esp-01 with the sketch, but it seems that I forgot how to do it because I was unable to upload the sketch to the chip!
> 
> I'll try again in the next few days

I can't even tell you how many hours I spent on the board :). I can follow tutorials and guides but my electronics knowledge is very limited, so it takes me forever to troubleshoot something.

**Comment by arkypita** [comment]
I was able to flash esp-01 (tested with blink sketch) then with telnet sketch.

Unlucky cannot find any wifi connection created by esp. Try with new and older version of library (both esp board library and wifimanager downgrader and upgraded). For me is a mistery.

More test next days.

Please keep in mind that esp is current hungry device, do not power up the board with usb only. Plug power supply.
Board like yours need proper different dip-switch configuration for flash and for run. Check board documentation.

Based on lgrbl telemetry 86 user use LaserWebESP8266, 50 use Telnet. Not so much but prove it can work.
Keep trying

**Comment by ssombra** [comment]
> I was able to flash esp-01 (tested with blink sketch) then with telnet sketch.
> 
> Unlucky cannot find any wifi connection created by esp. Try with new and older version of library (both esp board library and wifimanager downgrader and upgraded). For me is a mistery.
> 
> More test next days.
> 
> Please keep in mind that esp is current hungry device, do not power up the board with usb only. Plug power supply.
> Board like yours need proper different dip-switch configuration for flash and for run. Check board documentation.
> 
> Based on lgrbl telemetry 86 user use LaserWebESP8266, 50 use Telnet. Not so much but prove it can work.
> Keep trying

The same happens to me, I can flash the board but then I can not find the Access point created by the ESP or, in some very rare cases I see the Access Point for a while (then it disappears) but I can't connect to it. I actually tried with an external power supply (12v 2A) and it is the same. I just tried flashing the telnet sketch to a ESP-01s module and it works as expected, so the problem should be with the UNO + ESP8266 board itself. I still would love to use that board, so I am looking forward for any development you could get. 

By the way, I connected the ESP-01s module directly to an UNO board without resistors or voltage dividers (taking advantage of the 3.3v pin in the UNO): RX to TX, TX to RX, Reset to RST, 3.3v to 3.3v and GND to GND and it works well. What are the resistors connected to RX-TX in your schematics for?

**Comment by arkypita** [comment]
> What are the resistors connected to RX-TX in your schematics for?

Arduino is a 5v device, esp8266 is limited to 3.3v.
Resistor act as voltage divider to protect esp rx pin.

**Comment by ssombra** [comment]
> > What are the resistors connected to RX-TX in your schematics for?
> 
> Arduino is a 5v device, esp8266 is limited to 3.3v.
> Resistor act as voltage divider to protect esp rx pin.

I see. thanks.

**Comment by gmmanonymus111** [comment]
## How to program Uno r3+wifi microcontroller for LaserGRBL

#### 1. ESP8266 programming with WiFiAccessPoint example sketch _(optional, to test WiFi)_

- First, make sure your switches are set to the following:

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | OFF  | OFF  | OFF  | OFF  | ON  | ON  | ON  | -  |

This way it is set to programming mode, and the CH340 is connected to the ESP chip.

------------

I'm going to assume you have added ESP8266 boards to Arduino IDE (In case I'm wrong, there is a tutorial for that, near the end of this document).

- Once you have launched Arduino IDE, set your board type as `Generic ESP8266 Module` (`Tools` -> `Board` -> `ESP8266 boards` -> `Generic ESP8266 Module` in Arduino IDE 1.8.13).
![image](https://user-images.githubusercontent.com/9658243/97589058-63664880-19fd-11eb-8389-eb954f734a37.png)
- Set CPU Frequency to 80Mhz
- Set Crystal Frequency to 26Mhz (40Mhz caused errors for me, the board kept restarting)
- Set Upload Speed to 115200
- Set the correct COM port
- The rest can be set as seen on the image above

- Click `Upload` button. This should upload the sketch to the board
![image](https://user-images.githubusercontent.com/9658243/97590767-40d52f00-19ff-11eb-8397-d7190120cf12.png)
- Once done, set the switches as the following (#7 switch is set to OFF):

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | OFF  | OFF  | OFF  | OFF  | ON  | ON  | OFF  | -  |

This way CH340 is connected to the ESP chip, and you can see Serial Monitor output in Arduino IDE

- **Restart your board!**
- After a few seconds, a WiFi point should show up with the name `ESPap`
![image](https://user-images.githubusercontent.com/9658243/97591419-e5577100-19ff-11eb-8e38-6e0405815e84.png)
- You can connect to it, the default password is `thereisnospoon`
- Once connected, you can check your connection by typing `192.168.1.4` in your favourite browser's address bar. You should see a similar webpage:
![image](https://user-images.githubusercontent.com/9658243/97591864-5434ca00-1a00-11eb-8b7e-9375b6d0ca12.png)
- If eveything is working, you can continue to flash [ESP-LINK](https://github.com/jeelabs/esp-link "ESP-LINK")

------------

------------

#### 2. ESP8266 programming with [ESP-LINK](https://github.com/jeelabs/esp-link "ESP-LINK") and [ESPTool](https://github.com/espressif/esptool "ESPTool")

- First, make sure your switches are set to the following:

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | OFF  | OFF  | OFF  | OFF  | ON  | ON  | ON  | -  |

This way it is set to programming mode, and the CH340 is connected to the ESP chip.

------------

Again, I'm going to assume you have downloaded ESPTool (In case I'm wrong, there is a tutorial for that, near the end of this document).

- First, download [the latest ESP-LINK release](https://github.com/jeelabs/esp-link/releases/latest "ESP-LINK")
- Once downloaded, extarct it to somewhere. I'm goin to use esp8266 folder on my desktop
- Open Command Prompt in the folder you extracted the files into.
(TL:DR one-liner to flash eveything: `esptool.py --port COM4 write_flash 0x00000 boot_v1.6.bin 0x3FE000 blank.bin 0x3FC000 esp_init_data_default.bin 0x01000 user1.bin`)
- First, you have to flash a bootloader. I'll use the one included the archive. The command should be `esptool.py --port COM4 write_flash 0x00000 boot_v1.6.bin`
**Make sure to set the correct COM port!** If you get an error as `PermissionDenied` while uploading, make sure you have closed the Serial Monitor in Arduino IDE
![image](https://user-images.githubusercontent.com/9658243/97596192-c1e2f500-1a04-11eb-8f25-d535d5caf802.png)
After it was closed:
![image](https://user-images.githubusercontent.com/9658243/97595728-45501680-1a04-11eb-9011-0f0ae2b5e5a7.png)
- **Reset your board!**
- Now flash the blank space. The command: `esptool.py --port COM4 write_flash 0x3FE000 blank.bin`
The output should be similar to the above picture.
- **Reset your board!**
- Now flash the init_data binary: `esptool.py --port COM4 write_flash 0x3FC000 esp_init_data_default.bin`
- **Reset your board!**
- Finally, flash the user binary: `esptool.py --port COM4 write_flash 0x01000 user1.bin`
This might take a little longer. Once finished, you should set the switches as the following(#7 switch is set to OFF):

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | OFF  | OFF  | OFF  | OFF  | ON  | ON  | OFF  | -  |

This way CH340 is connected to the ESP chip, and you can see Serial Monitor output in Arduino IDE
- **Reset your board!**
- When your board comes online, you should see a new Access Point, like `ESP_6608C0`
![image](https://user-images.githubusercontent.com/9658243/97597785-83e6d080-1a06-11eb-89d6-12ed68655f17.png)
- Connect to it, then type `192.168.4.1` in your browser's address bar, you should get something like this:
![image](https://user-images.githubusercontent.com/9658243/97598479-50587600-1a07-11eb-976e-f6a649305cb9.png)
- Select `WiFi Station` from the menu on the left side of the webpage.
- Click on `Switch to STA+AP mode to scan.` at the `WiFi Association` panel to scan for WiFi networks to connect to.
- Select your network, type in your password (if there is one), then click connect
![image](https://user-images.githubusercontent.com/9658243/97599701-8f3afb80-1a08-11eb-9358-9ce8df69cf3e.png)
- It might error out, but it is connected to the network. To make sure, check the serial monitor in Arduino IDE after a restart.
![image](https://user-images.githubusercontent.com/9658243/97599947-d6c18780-1a08-11eb-9bcc-b59dbc1b4541.png)

-------

-------

#### 3. ATmega328P programming with grblUpload "example" sketch

- First, make sure your switches are set to the following:

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | OFF  | OFF  | ON  | ON  | OFF  | OFF  | OFF  | -  |

This way the CH340 is connected to the ATmega (Arduino Uno) chip.
The uploading and monitoring is achieved by the same switch states. (You don't need to "switch & reset" to change from programming mode to monitoring mode)

--------

I'm going to assume you have set up grbl in your Arduino IDE. I'm not going to write a tutorial here, as ther are plenty how to do it.

- First, set your board to Arduino Uno. In Arduino IDE 1.8.13 it is `Tools` -> `Board` -> `Arduino AVR Boards` -> `Arduino Uno`
![image](https://user-images.githubusercontent.com/9658243/97601860-f0fc6500-1a0a-11eb-88f7-4bb28bc2f8ee.png)
- Set COM port
- Open grblUpload sketch from `File` ->`Examples` -> `grbl` -> `grblUpload`
![image](https://user-images.githubusercontent.com/9658243/97602417-8ef02f80-1a0b-11eb-9c61-11fc7ed3a1ab.png)
- Click the `Upload` button
When finished, Serial monitor should show something like this:
```
Grbl 1.1h ['$' for help]
[MSG:'$H'|'$X' to unlock]
```

-------

-------

#### 4. Connecting ATmega328P with ESP8266
- You only need to set the switches as the following:

| Switch | 1  |  2 | 3  | 4  | 5  | 6  | 7  | 8  |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| **State** | ON  | ON  | OFF  | OFF  | OFF  | OFF  | OFF  | -  |

- When done, ESP-link should show you GRBL command responses while you are accessing it through your browser, or through telnet, or through LaserGRBL
![image](https://user-images.githubusercontent.com/9658243/97604292-99abc400-1a0d-11eb-8188-1af5e41be80c.png)

-------

-------

#### 5. Bonus: installing ESP boards and ESPTool

- For ESP8266 boards, see [ESP8266 installing with boards manager](https://github.com/esp8266/Arduino#installing-with-boards-manager "ESP_Arduino")
- For ESPTool, it is just a simple Python command: `pip install esptool`. You can download Python from [here](https://www.python.org/downloads/ "here")

**Comment by ssombra** [comment]
Thanks for taking the time for this very comprehensive tutorial. I followed it and I get stuck in the same place, even with the WiFiAccessPoint example:

![image](https://user-images.githubusercontent.com/42944253/97626777-fac49f00-19ef-11eb-97f9-5c9af5f01636.png)

The sketch uploads, the Access Point shows in the network list, but then I can´t connect to it. I didn´t go to the ESP-link / Esptool part because I keep getting the same no connection error. I bought two of these boards and both have the same problem, I know that it is very unlikely, but maybe both are defective, who knows. 

Something I noticed which is very strange is that the access point created by the ESP never shows in my phone (iphone), only in the PC and not always all the time.

**Comment by gmmanonymus111** [comment]
Can you connect to the ESPap using any other device (e.g. a smartphone)
A user mentioned Intel driver issues at [ESPDeauther](https://github.com/SpacehuhnTech/esp8266_deauther/issues/754), but they were able to connect to the AP with a smartphone
From here on, I'm only guessing :/

**Comment by ssombra** [comment]
Actually I had edited my previous comment to say that the access point doesn't even appear in my iPhone, not a single time from all the tests I made.

**Comment by gmmanonymus111** [comment]
Oh, didn't see that. Found [another issue](https://github.com/esp8266/Arduino/issues/1094), where they explain that esp8266 has only one radio, and it gets bugged a lot because of cached settings, it can be helped by calling `ESP.eraseConfig();`, `WiFi.disconnect(true);` and `WiFi.mode(WIFI_AP);`

**Comment by ssombra** [comment]
Thanks for the links, I checked them and tried all the ideas there (eraseConfig, disconnect, changing phy mode) and none worked. Will keep looking for info.

**Comment by harryberlin** [comment]
i think wifi connection has lot of risk.
is there a watchdog to stop and hold the system, if connection is lost?

**Comment by gmmanonymus111** [comment]
Not yet, but I think WiFi connection is very stable (for me, there were no accidental disconnects at all), and for a small machine like this, the biggest problem could be you burn a big black dot, if the laser stays on, and you didn't hear the disconnect sound. For bigger, more serious machines, it is better to use wired connection of course.

**Comment by arkypita** [comment]
@harryberlin the same risk you have with USB.
For what I see only Ortur has implemented a sort of watchdog in their firmware that shut down the laser if USB is disconnected. Standard grbl machine doesn't do anything.

Moreover WiFi allow you put the engraver outside, far from PC where nothing can burn, instead inside you home on the same table you put your PC.
